File: pkg/tracing/tracing.go
package tracing

import (
	"context"
	"os"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/jaeger"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
	"go.opentelemetry.io/otel/trace"
)

func InitTracer(serviceName string) (func(context.Context) error, error) {
	jaegerURL := getEnv("JAEGER_ENDPOINT", "http://localhost:14268/api/traces")
	exp, err := jaeger.New(jaeger.WithCollectorEndpoint(jaeger.WithEndpoint(jaegerURL)))
	if err != nil {
		return nil, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exp),
		sdktrace.WithResource(resource.NewWithAttributes(
			semconv.SchemaURL,
			semconv.ServiceName(serviceName),
			semconv.ServiceVersion(getEnv("APP_VERSION", "dev")),
		)),
	)

	otel.SetTracerProvider(tp)
	return tp.Shutdown, nil
}

func GetTracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

File: src/python/utils/tracing.py
"""OpenTelemetry distributed tracing for Python services."""

import os
from opentelemetry import trace
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


def init_tracer(service_name: str) -> trace.Tracer:
    """
    Initialize OpenTelemetry tracing.
    
    Args:
        service_name: Name of the service
        
    Returns:
        Tracer instance
    """
    # Create resource
    resource = Resource(attributes={
        SERVICE_NAME: service_name,
        "service.version": os.getenv("APP_VERSION", "dev"),
    })
    
    # Create Jaeger exporter
    jaeger_endpoint = os.getenv("JAEGER_ENDPOINT", "localhost")
    jaeger_port = int(os.getenv("JAEGER_PORT", "6831"))
    
    jaeger_exporter = JaegerExporter(
        agent_host_name=jaeger_endpoint,
        agent_port=jaeger_port,
    )
    
    # Create trace provider
    provider = TracerProvider(resource=resource)
    processor = BatchSpanProcessor(jaeger_exporter)
    provider.add_span_processor(processor)
    
    # Set global trace provider
    trace.set_tracer_provider(provider)
    
    return trace.get_tracer(service_name)


# Example usage:
#
# from src.python.utils.tracing import init_tracer
#
# tracer = init_tracer("trainer")
#
# with tracer.start_as_current_span("training_batch"):
#     # Your training code here
#     pass
